#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/bool.hpp"
#include "std_msgs/msg/int64.hpp"
#include <lgpio.h>
#include <functional>
#include <memory>
using namespace std::chrono_literals;

#define LFLAGS 0

namespace coral_cam{
    class Gpio : public rclcpp::Node{
        public:
            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Constructors and Destructors                                                                          //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////
            Gpio(const rclcpp::NodeOptions &options);

            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Public Methods                                                                                        //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////
            
        private:
            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Callbacks                                                                                             //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////
            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Private Methods                                                                                       //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////

            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Node,Timers, Publishers and Subscribers                                                                      //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////
            rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr gpioHandlePublisher_;
            rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr buttonPinPublisher_;
            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Constants                                                                                             //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////

            ////////////////////////////////////////////////////////////////////////////////////////////////////////////
            // Variables                                                                                             //
            //////////////////////////////////////////////////////////////////////////////////////////////////////////
            int gpioHandle_;
            int gpioNumber_;
            int buttonPin_;

            std_msgs::msg::Int64 gpioHandleMessage_;
            std_msgs::msg::Int64 buttonPinMessage_;

    };  
}